chapter/1/rpc-reviewed-dont-keep.md

```diff
@@ -0,0 +1,619 @@
+Researchers realized the limitations of RPC and focussed on rectifying and
+instead of enforcing RPC, they started to use RPC in applications where it was
+needed. The designer started adding exception-handling, async, network failure
+handling and heterogenity between different languages/devices to RPC.
+
+Perhaps, the earliest system in this era was SunRPC {% cite sunnfs --file rpc %}
+used for the Sun Network File System(NFS). This SunRPC has gone under various
+additions and is now referred to as Open Network Computing RPC(ONC RPC).
+
+Soon to follow SunRPC was the language-agnostic CORBA{% cite corba --file rpc %}
+which was followed by Java RMI{% cite rmipaper --file rpc %}. CORBA and RMI have
+also undergone various modifications as internet standards were set and TCP/IP
+became the norm.
+
+A new breed of RPC also started in this era(early 2000's), Async RPC, giving
+rise to systems that use *futures* and *promises*, like Finagle{% cite finagle
+--file rpc %} and Cap'n Proto(post-2010).
+
+In the post-2000 era, MAUI{% cite maui --file rpc %}, Cap'n Proto{% cite
+capnproto --file rpc %}, gRPC{% cite grpc --file rpc %}, Thrift{% cite thrift
+--file rpc %} and Finagle{% cite finagle --file rpc %} have been released, which
+have significantly boosted the widespread use of RPC. A level overview of some
+of the most important RPC implementation is as follows.
+
+#### Java Remote Method Invocation
+
+Java RMI (Java Remote Method Invocation){% cite rmibook --file rpc %} is a Java
+implementation for performing RPC (Remote Procedure Calls) between a client and
+a server.  The client using a stub passes via a socket connection the
+information over the network to the server.  The Remote Object Registry (ROR){%
+cite rmipaper --file rpc %} on the server contains the references to objects
+that can be accessed remotely and through which the client will connect to.  The
+client then can request of the invocation of methods on the server for
+processing the requested call and then responds with the answer.  RMI provides
+some security by being encoded but not encrypted, though that can be augmented
+by tunneling over a secure connection or other methods.
+
+#### CORBA
+
+CORBA (Common Object Request Broker Architecture){% cite corba --file rpc %} was
+created by the Object Management Group {% cite corbasite --file rpc %} to allow
+for language-agnostic communication among multiple computers.  It is an
+object-oriented model defined via an Interface Definition Language (IDL) and the
+communication is managed through an Object Request Broker (ORB).  Each client
+and server have an ORB by which they communicate.  The benefits of CORBA is that
+it allows for multi-language implementations that can communicate with each
+other, but much of the criticism around CORBA relates to poor consistency among
+implementations.
+
+#### XML-RPC and SOAP
+
+SOAP (Simple Object Access Protocol) is a successor of XML-RPC as a web-services
```

Comment:
@MuzammilAR For XML-RPC you can mention the following:

### XML-RPC

The XML-RPC specifications {% cite Wiener %} performs an HTTP Post to a server formatted as XML composed of a `Header`, `Payload` that calls only one method.  Below is an example {% cite Wiener %}:

```XML

POST /RPC2 HTTP/1.0
User-Agent: Frontier/5.1.2 (WinNT)
Host: betty.userland.com
Content-Type: text/xml
Content-length: 181

<?xml version="1.0"?>
<methodCall>
   <methodName>examples.getStateName</methodName>
   <params>
      <param>
         <value><i4>41</i4></value>
         </param>
      </params>
   </methodCall>
```

The header has to provide the `User-Agent`, `Host`, `Content-Type` as `text/xml` and `Content-Length`.

The `Payload` has to initiate a `methodCall` structure by specifying the name via `methodName` and associated parameter values.  Parameters for the method can be scalar, structures or (recursive) arrays.  The types of scalar can be one of `i4`, `int`, `boolean`, `string`, `double`, `dateTime.iso8601` or `base64`.

Structures will have members containing at least one name and value as such {% cite Wiener %}:

```XML
<struct>
   <member>
      <name>lowerBound</name>
      <value><i4>18</i4></value>
      </member>
   <member>
      <name>upperBound</name>
      <value><i4>139</i4></value>
      </member>
   </struct>
```

Arrays will have a data element with multiple values as such {% cite Wiener %}:

```XML
<array>
   <data>
      <value><i4>12</i4></value>
      <value><string>Egypt</string></value>
      <value><boolean>0</boolean></value>
      <value><i4>-31</i4></value>
      </data>
   </array>
```

The response to a request will have the `methodResponse` with `params` and values, or a `fault` with the associated `faultCode` in case of an error {% cite Wiener %}:

```XML
HTTP/1.1 200 OK
Connection: close
Content-Length: 158
Content-Type: text/xml
Date: Fri, 17 Jul 1998 19:55:08 GMT
Server: UserLand Frontier/5.1.2-WinNT

<?xml version="1.0"?>
<methodResponse>
   <params>
      <param>
         <value><string>South Dakota</string></value>
         </param>
      </params>
   </methodResponse>
```

### Reference

[Wiener] http://xmlrpc.scripting.com/spec.html